include/tvm/ir/diagnostic.h

```diff
@@ -0,0 +1,254 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file diagnostic.h
+ * \brief A new diagnostic interface for TVM error reporting.
+ *
+ * A prototype of the new diagnostic reporting interface for TVM.
+ *
+ * Eventually we hope to promote this file to the top-level and
+ * replace the existing errors.h.
+ */
+
+#ifndef TVM_IR_DIAGNOSTIC_H_
+#define TVM_IR_DIAGNOSTIC_H_
+
+#include <tvm/ir/span.h>
+#include <tvm/ir/module.h>
+#include <tvm/parser/source_map.h>
+#include <tvm/runtime/container.h>
+#include <tvm/runtime/object.h>
+#include <tvm/support/logging.h>
+
+#include <fstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace tvm {
+
+using tvm::parser::SourceMap;
+
+static const char* kTVM_INTERNAL_ERROR_MESSAGE = "An internal invariant was violated during the execution of TVM" \
+                                          "please read TVM's error reporting guidelines at discuss.tvm.ai/thread";
+
+static const char* kINDENT = "  ";
+
+#define ICHECK_BINARY_OP(name, op, x, y)                              \
+  if (dmlc::LogCheckError _check_err = dmlc::LogCheck##name(x, y))    \
+    dmlc::LogMessageFatal(__FILE__, __LINE__).stream()                \
+      << kTVM_INTERNAL_ERROR_MESSAGE << std::endl                     \
+      << kIndent << "Check failed: " << #x " " #op " " #y << *(_check_err.str) << ": "
+
+#define ICHECK(x)                                          \
+  if (!(x))                                                \
+    dmlc::LogMessageFatal(__FILE__, __LINE__).stream()     \
+      << kTVM_INTERNAL_ERROR_MESSAGE                       \
+      << kINDENT << "Check failed: " #x << ": "
+
+#define ICHECK_LT(x, y) ICHECK_BINARY_OP(_LT, <, x, y)
+#define ICHECK_GT(x, y) ICHECK_BINARY_OP(_GT, >, x, y)
+#define ICHECK_LE(x, y) ICHECK_BINARY_OP(_LE, <=, x, y)
+#define ICHECK_GE(x, y) ICHECK_BINARY_OP(_GE, >=, x, y)
+#define ICHECK_EQ(x, y) ICHECK_BINARY_OP(_EQ, ==, x, y)
+#define ICHECK_NE(x, y) ICHECK_BINARY_OP(_NE, !=, x, y)
+#define ICHECK_NOTNULL(x) \
+  ((x) == NULL ? dmlc::LogMessageFatal(__FILE__, __LINE__).stream() \
+  << kTVM_INTERNAL_ERROR_MESSAGE                                    \
+  << kINDENT << "Check not null: "  #x << ' ', (x) : (x)) // NOLINT(*)
+
+/*! \brief The diagnostic level, controls the printing of the message. */
+enum class DiagnosticLevel {
+  kBug,
+  kError,
+  kWarning,
+  kNote,
+  kHelp,
+};
```

Comment:
Maybe using exact integers?

```suggestion
enum class DiagnosticLevel : int {
  kBug = 50,
  kError = 40,
  kWarning = 30,
  kNote = 20,
  kHelp = 10,
};
```

Reply:
Yeah this probably useful for Python I generally like to avoid using int enums.